Sources/CryptoSwift/RSA/RSA.swift

```diff
@@ -0,0 +1,383 @@
+//
+//  CryptoSwift
+//
+//  Copyright (C) 2014-2021 Marcin Krzyżanowski <marcin@krzyzanowskim.com>
+//  This software is provided 'as-is', without any express or implied warranty.
+//
+//  In no event will the authors be held liable for any damages arising from the use of this software.
+//
+//  Permission is granted to anyone to use this software for any purpose,including commercial applications, and to alter it and redistribute it freely, subject to the following restrictions:
+//
+//  - The origin of this software must not be misrepresented; you must not claim that you wrote the original software. If you use this software in a product, an acknowledgment in the product documentation is required.
+//  - Altered source versions must be plainly marked as such, and must not be misrepresented as being the original software.
+//  - This notice may not be removed or altered from any source or binary distribution.
+//
+
+// Foundation is required for `Data` to be found
+import Foundation
+
+// Note: The `BigUInt` struct was copied from:
+// https://github.com/attaswift/BigInt
+// It allows fast calculation for RSA big numbers
+
+public final class RSA: DERCodable {
+  /// RSA Object Identifier Bytes (rsaEncryption)
+  static var primaryObjectIdentifier: Array<UInt8> = Array<UInt8>(arrayLiteral: 0x2A, 0x86, 0x48, 0x86, 0xF7, 0x0D, 0x01, 0x01, 0x01)
```

Comment:
Ah, thanks for pointing that out. I was working on PEM support simultaneously	 and the `primary` and `secondary` PKCS#8 OID’s are used when constructing a PEM formatted representation of the key. Although it’s not necessary for the Raw External DER Representation of a key, If the DERCodable protocol enforces these OID’s then we can extend the protocol with PEM functionality. Here’s an example of what I had in mind…
``` swift
public extension DEREncodable {

  internal func exportPublicKeyPEMRaw() throws -> Array<UInt8> {
    let publicDER = try self.publicKeyDER()

    let node:ASN1.Node = .sequence(nodes: [
        .sequence(nodes: [
          .objectIdentifier(data: Data(Self.primaryObjectIdentifier)),
          .null
        ]),
        .bitString(data: Data( publicDER ))
      ])
      
    return ASN1.Encoder.encode(node)
  }

  func exportPublicKeyPEM(withHeaderAndFooter:Bool = true) throws -> Array<UInt8> {
    let base64String = try self.exportPublicKeyPEMRaw().toBase64()
    let bodyString = base64String.chunks(ofCount: 64).joined(separator: "\n")
    let bodyUTF8Bytes = bodyString.bytes
    
    if withHeaderAndFooter {
      let header = PEM.PEMType.publicKey.headerBytes + [0x0a]
      let footer = [0x0a] + PEM.PEMType.publicKey.footerBytes
    
      return header + bodyUTF8Bytes + footer
    } else {
      return bodyUTF8Bytes
    }
  }
```

But maybe it makes more sense to define a PEM protocol in the future that extends DERCodable and enforces the PKCS8 OID’s. Something like…
``` swift
protocol PEM:DERCodable {
  /// The keys ASN1 primary object identifier (ex: RSA --> rsaEncryption --> [0x2A, 0x86, 0x48, 0x86, 0xF7, 0x0D, 0x01, 0x01, 0x01])
  static var primaryObjectIdentifier:Array<UInt8> { get }
  /// The keys ASN1 secondary object identifier (ex: RSA --> .null)
  static var secondaryObjectIdentifier:Array<UInt8>? { get }
}

extension PEM {
  /// Instantiates a DERDecodable Key from a PEM string
  /// - Parameters:
  ///   - pem: The PEM file to import
  ///   - password: A password to use to decrypt an encrypted PEM file
  ///   - asType: The underlying DERDecodable Key Type (ex: RSA.self)
  init<Key:DERDecodable>(pem: String, password: String? = nil, asType:Key.Type = Key.self) throws {
      …
  }

  /// PublicKey PEM Export Functions
  func exportPublicKeyPEM(withHeaderAndFooter:Bool) throws -> Array<UInt8> { … }
  func exportPublicKeyPEMString(withHeaderAndFooter:Bool) throws -> String { … }
  
  /// PrivateKey PEM Export Functions
  func exportPrivateKeyPEM(withHeaderAndFooter:Bool) throws -> Array<UInt8> { … }
  func exportPrivateKeyPEMString(withHeaderAndFooter:Bool) throws -> String { .. }
}
```
If you’d prefer the later, then I can go ahead and remove the `primary` and `secondary` OID’s from `RSA` and both of the `DEREncodable` and `DERDecodable` protocols.
 
Let me know which you prefer (I think I prefer the later, especially since there might be DERCodable objects that don't require / have a primary and secondary object id).

Reply:
I think that the second approach is better, because not all DERCodable objects require PEM support. And also I think that `PEMCodable` would be more consistant for the name of the protocol.